h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstScale.java

```diff
@@ -35,30 +35,69 @@ public String str() {
 
   @Override
   public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
-    Frame fr = stk.track(asts[1].exec(env)).getFrame();
-    int ncols = fr.numCols();
+    final Frame originalFrame = stk.track(asts[1].exec(env)).getFrame();
+    final Frame numericFrame = new Frame(); // filter the frame to only numerical columns
+    for (int i = 0; i < originalFrame.numCols(); i++) {
+      Vec v = originalFrame.vec(i);
+      if (v.get_type() == Vec.T_NUM) {
+        numericFrame.add(originalFrame.name(i), v);
+      }
+    }
+
+    final double[] means = calcMeans(env, asts[2], numericFrame);
+    final double[] mults = calcMults(env, asts[3], numericFrame);
+    // Update in-place.
+    new InPlaceSlateTask(means, mults).doAll(numericFrame);
+
+    // Return the original frame (with categorical and other columns), this is okay to do because the update was made in-place.
+    return new ValFrame(originalFrame);
+  }
+
+  private static class InPlaceSlateTask extends MRTask<InPlaceSlateTask> {
+    private final double[] _means;
+    private final double[] _mults;
+
+    InPlaceSlateTask(double[] means, double[] mults) {
+      _means = means;
+      _mults = mults;
+    }
 
-    // Peel out the bias/shift/mean
+    @Override
+    public void map(Chunk[] cs) {
+      for (int i = 0; i < cs.length; i++)
+        for (int row = 0; row < cs[i]._len; row++)
+          cs[i].set(row, (cs[i].atd(row) - _means[i]) * _mults[i]);
+    }
+
+  }
+  
+  // Peel out the bias/shift/mean
+  private static double[] calcMeans(Env env, AstRoot meanSpec, Frame fr) {
+    final int ncols = fr.numCols();
     double[] means;
-    if (asts[2] instanceof AstNumList) {
-      means = ((AstNumList) asts[2]).expand();
+    if (meanSpec instanceof AstNumList) {
+      means = ((AstNumList) meanSpec).expand();
       if (means.length != ncols)
-        throw new IllegalArgumentException("Numlist must be the same length as the columns of the Frame");
+        throw new IllegalArgumentException("Numlist must be the same length as the numeric columns of the Frame");
```

Comment:
I'd personally test if the exception is thrown in JUnit.

Reply:
Sure, thanks!